ragna/core/_rag.py

```diff
@@ -220,17 +223,20 @@ async def answer(self, prompt: str, *, stream: bool = False) -> Message:
                 detail=RagnaException.EVENT,
             )
 
-        self._messages.append(Message(content=prompt, role=MessageRole.USER))
-
         sources = await self._run(self.source_storage.retrieve, self.documents, prompt)
 
+        question = Message(content=prompt, role=MessageRole.USER, sources=sources)
+        self._messages.append(question)
+
         answer = Message(
-            content=self._run_gen(self.assistant.answer, prompt, sources),
+            content=self._run_gen(self.assistant.answer, self._messages),
             role=MessageRole.ASSISTANT,
             sources=sources,
         )
-        if not stream:
```

Comment:
I was struggling to get streaming to work with the list of messages. Either I would get an error like this when the answer's Message content is still an AsyncIterable
```
    | Traceback (most recent call last):
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/starlette/responses.py", line 261, in wrap
    |     await func()
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/starlette/responses.py", line 250, in stream_response
    |     async for chunk in self.body_iterator:
    |   File "/home/blake/code/ragna/ragna/deploy/_api/core.py", line 328, in to_jsonl
    |     async for model in models:
    |   File "/home/blake/code/ragna/ragna/deploy/_api/core.py", line 302, in message_chunks
    |     content_chunk = await anext(core_answer_stream)
    |   File "/home/blake/code/ragna/ragna/_compat.py", line 85, in anext_impl
    |     return await ait.__anext__()
    |   File "/home/blake/code/ragna/ragna/core/_components.py", line 202, in __aiter__
    |     async for chunk in self._content_stream:
    |   File "/home/blake/code/ragna/ragna/core/_rag.py", line 403, in _run_gen
    |     async for item in async_gen:
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/starlette/concurrency.py", line 65, in iterate_in_threadpool
    |     yield await anyio.to_thread.run_sync(_next, as_iterator)
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/anyio/to_thread.py", line 56, in run_sync
    |     return await get_async_backend().run_sync_in_worker_thread(
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 2177, in run_sync_in_worker_thread
    |     return await future
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 859, in run
    |     result = context.run(func, *args)
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/starlette/concurrency.py", line 54, in _next
    |     return next(iterator)
    |   File "/home/blake/code/ragna/ragna/assistants/_demo.py", line 25, in answer
    |     yield self._default_answer(messages)
    |   File "/home/blake/code/ragna/ragna/assistants/_demo.py", line 39, in _default_answer
    |     prompt = messages[-1].content.strip()
    |   File "/home/blake/code/ragna/ragna/core/_components.py", line 219, in content
    |     raise RuntimeError(
    | RuntimeError: Message content cannot be accessed without having iterated over it, e.g. `async for chunk in message`, or reading the content, e.g. `await message.read()`, first.
```
or I anger the async gods further by trying to convert the message `.content` into a string by iterating over it in advance:
```
   | Traceback (most recent call last):
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/starlette/responses.py", line 261, in wrap
    |     await func()
    |   File "/home/blake/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/starlette/responses.py", line 250, in stream_response
    |     async for chunk in self.body_iterator:
    |   File "/home/blake/code/ragna/ragna/deploy/_api/core.py", line 328, in to_jsonl
    |     async for model in models:
    |   File "/home/blake/code/ragna/ragna/deploy/_api/core.py", line 302, in message_chunks
    |     content_chunk = await anext(core_answer_stream)
    |   File "/home/blake/code/ragna/ragna/_compat.py", line 85, in anext_impl
    |     return await ait.__anext__()
    |   File "/home/blake/code/ragna/ragna/core/_components.py", line 202, in __aiter__
    |     async for chunk in self._content_stream:
    |   File "/home/blake/code/ragna/ragna/core/_rag.py", line 403, in _run_gen
    |     async for item in async_gen:
    |   File "/home/blake/code/ragna/ragna/assistants/_openai.py", line 81, in answer
    |     formatted_messages = await self._format_message_sources(messages)
    |   File "/home/blake/code/ragna/ragna/assistants/_openai.py", line 35, in _format_message_sources
    |     await message.read()
    |   File "/home/blake/code/ragna/ragna/core/_components.py", line 212, in read
    |     async for _ in self:
    |   File "/home/blake/code/ragna/ragna/core/_components.py", line 202, in __aiter__
    |     async for chunk in self._content_stream:
    | RuntimeError: anext(): asynchronous generator is already running
```

Reply:
Changing the signature of `Assistant.answer` should have no effect. Could you please post a snippet to reproduce this?

Reply:
The issue is with trying to use `message.content` in `Assistant._format_message_sources`. Some of the Messages in the message list are still `AsyncIterable`s. If I try to call `.read()` on a message within the assistant I get
```
RuntimeError: anext(): asynchronous generator is already running
```
Here's an example of that:
```python
def Assistant:
    ...
    async def _format_message_sources(self, messages: list[Message]) -> str:
        sources_prompt = "Include the following sources in your answer:"
        formatted_messages = []
        for message in messages:
            content = await message.read()
            if message.role == MessageRole.USER:
                formatted_messages.append(
                    {
                        "content": sources_prompt
                        + "\n\n".join(source.content for source in message.sources),
                        "role": MessageRole.SYSTEM,
                    }
                )

            formatted_messages.append({"content": content, "role": message.role})
        return formatted_messages

    async def answer(
        self,
        messages: list[Message] = [],
        *,
        max_new_tokens: int = 256,
    ) -> AsyncIterator[str]:
        formatted_messages = await self._format_message_sources(messages)
        async for data in self._stream(
            formatted_messages, max_new_tokens=max_new_tokens
        ):
            choice = data["choices"][0]
            if choice["finish_reason"] is not None:
                break

            yield cast(str, choice["delta"]["content"])
```
Then running `chat.answer` will produce the error when streaming is true. I just pushed a commit that demos this behavior.

Reply:
Without having it reproduced, I think the error comes from the fact that you want to use the old assistant message to generate a new one, but have never accessed it. This is not a standard use case as we can safely assume that a user would consume an answer before asking a follow-up question. And after a message once has been `.read()`, you can safely access `.content`.

Reply:
That makes sense to me and is how I'd expect this to behave. I experience this problem after asking a single question, so it's a mystery why `messages` contains an unread answer ever. I'll dig a little more.